ciStages.go

```diff
@@ -140,101 +146,35 @@ func runCIStages(ciCdRequest *helper.CiCdTriggerEvent) (artifactUploaded bool, e
 	if len(ciCdRequest.CommonWorkflowRequest.PreCiSteps) > 0 {
-		if !buildSkipEnabled {
-			util.LogStage("running PRE-CI steps")
-		}
-		// run pre artifact processing
-		preeCiStageOutVariable, step, err = RunCiCdSteps(STEP_TYPE_PRE, ciCdRequest.CommonWorkflowRequest.PreCiSteps, refStageMap, scriptEnvs, nil)
-		preCiDuration = time.Since(start).Seconds()
-		if err != nil {
-			log.Println(err)
-			return sendFailureNotification(string(PreCi)+step.Name, ciCdRequest.CommonWorkflowRequest, "", "", metrics, artifactUploaded, err)
-
-		}
-		// considering pull images from Container repo Plugin in Pre ci steps only.
-		// making it non-blocking if results are not available (in case of err)
-		resultsFromPlugin, err = extractOutResultsIfExists()
+		resultsFromPlugin, preCiStageOutVariable, err = runPreCiSteps(ciCdRequest, &metrics, buildSkipEnabled, refStageMap, scriptEnvs, artifactUploaded)
 		if err != nil {
-			log.Println("error in getting results", "err", err.Error())
+			return artifactUploaded, err
 		}
 	}
-	metrics.PreCiDuration = preCiDuration
 	var dest string
+	var digest string
 	if !buildSkipEnabled {
-		util.LogStage("Build")
-		// build
-		start = time.Now()
-		metrics.BuildStartTime = start
-		dest, err = helper.BuildArtifact(ciCdRequest.CommonWorkflowRequest) //TODO make it skipable
-		metrics.BuildDuration = time.Since(start).Seconds()
+		dest, digest, err = getImageDestAndDigest(ciCdRequest, metrics, scriptEnvs, refStageMap, preCiStageOutVariable, artifactUploaded)
 		if err != nil {
-			// code-block starts : run post-ci which are enabled to run on ci fail
-			postCiStepsToTriggerOnCiFail := getPostCiStepToRunOnCiFail(ciCdRequest.CommonWorkflowRequest.PostCiSteps)
-			if len(postCiStepsToTriggerOnCiFail) > 0 {
-				util.LogStage("Running POST-CI steps which are enabled to RUN even on CI FAIL")
-				// build success will always be false
-				scriptEnvs[util.ENV_VARIABLE_BUILD_SUCCESS] = "false"
-				// run post artifact processing
-				RunCiCdSteps(STEP_TYPE_POST, postCiStepsToTriggerOnCiFail, refStageMap, scriptEnvs, preeCiStageOutVariable)
-			}
-			// code-block ends
-			return sendFailureNotification(string(Build), ciCdRequest.CommonWorkflowRequest, "", "", metrics, artifactUploaded, err)
+			return artifactUploaded, err
 		}
-		log.Println(util.DEVTRON, " /Build")
 	}
 	var postCiDuration float64
 	start = time.Now()
 	metrics.PostCiStartTime = start
 	if len(ciCdRequest.CommonWorkflowRequest.PostCiSteps) > 0 {
-		util.LogStage("running POST-CI steps")
-		// sending build success as true always as post-ci triggers only if ci gets success
-		scriptEnvs[util.ENV_VARIABLE_BUILD_SUCCESS] = "true"
-		// run post artifact processing
-		_, step, err = RunCiCdSteps(STEP_TYPE_POST, ciCdRequest.CommonWorkflowRequest.PostCiSteps, refStageMap, scriptEnvs, preeCiStageOutVariable)
+		err = runPostCiSteps(ciCdRequest, scriptEnvs, refStageMap, preCiStageOutVariable, metrics, artifactUploaded, dest, digest)
```

Comment:
we are passing metrics by value here and above, are they going to be manipulated separately? if not please fix